Content.Server/TapeRecorder/TapeRecorderSystem.cs

```diff
@@ -0,0 +1,338 @@
+
+        /// <summary>
+        /// Flushes recorded buffer to tape memory. Overwrites data if the timestamps overlap
+        /// </summary>
+        private static void FlushBufferToMemory(TapeRecorderComponent component)
+        {
+            if (component.InsertedTape == null)
+                return;
+
+            component.InsertedTape.TimeStamp = (component.AccumulatedTime - component.RecordingStartTime);
+
+            //Clear the recorded messages between the start and end of our recording timestamps, since we're overwriting this part of the tape
+            component.InsertedTape.RecordedMessages.RemoveAll(x => x.MessageTimeStamp > component.RecordingStartTimestamp && x.MessageTimeStamp < component.InsertedTape.TimeStamp);
+            component.InsertedTape.RecordedMessages.AddRange(component.RecordedMessageBuffer);
+
+            component.RecordedMessageBuffer.Clear();
+
+            //sort the list by timestamp
+            component.InsertedTape.RecordedMessages.Sort((x, y) => x.MessageTimeStamp.CompareTo(y.MessageTimeStamp));
+
+            }
+
+        private void UpdateAppearance(TapeRecorderComponent component)
+        {
+            if (!TryComp<AppearanceComponent>(component.Owner, out var appearance))
+                return;
+
+            if (!component.Enabled && component.InsertedTape != null)
+            {
+                _appearanceSystem.SetData(component.Owner, TapeRecorderVisuals.Status, TapeRecorderState.Idle);
+                return;
+            }
+
+            _appearanceSystem.SetData(component.Owner, TapeRecorderVisuals.Status, component.CurrentMode);
+        }
+
+        private void OnUseInHand(EntityUid uid, TapeRecorderComponent component, UseInHandEvent args)
+        {
+            args.Handled = true;
+
+            if (component.Enabled || component.InsertedTape == null)
+            {
+                StopTape(component);
+                return;
+            }
+
+            if (!component.Enabled)
+                component.Enabled = true;
+            UpdateAppearance(component);
+
+            switch (component.CurrentMode) //idk how else to do this
+            {
+                case TapeRecorderState.Play:
+                    StartPlaying(component);
+                    break;
+                case TapeRecorderState.Record:
+                    StartRecording(component);
+                    break;
+                case TapeRecorderState.Rewind:
+                    StartRewinding(component);
+                    break;
+                case TapeRecorderState.Idle:
+                    break;
+                case TapeRecorderState.Empty:
+                    component.Enabled = false;
+                    break;
+            }
+        }
+
+        /// <summary>
+        /// Records heard messages to the message buffer
+        /// </summary>
+        private void OnChatMessageHeard(EntityUid uid, TapeRecorderComponent component, ChatMessageHeardNearbyEvent args)
+        {
```

Comment:
The part that most concerns me about this is the lack of check for an obstruction. Right now on master, you can (clearly) hear chat through walls - but I'm not sure if this is meant to be intended. For example, if you look at something like HandheldRadio, there's an obstruction check there - however, that also uses the outdated IListen component in order to do the logic check, but it still does the check.

Reply:
I experimented with adding a check but then it'd have problems hearing things through glass windows and such, which you'd want to have especially for questioning someone in a jail cell or something like that. I think it'll have to wait until regular chat has obstruction checking

Reply:
Yeah - it's a concern that could probably be fixed up by addressing the root cause and making chat not go through walls